zingolib/src/wallet/transaction_record.rs

```diff
@@ -0,0 +1,343 @@
+    pub status: ConfirmationStatus,
+
+    // Timestamp of Tx. Added in v4
+    pub datetime: u64,
+
+    // Txid of this transaction. It's duplicated here (It is also the Key in the HashMap that points to this
+    // WalletTx in LightWallet::txs)
+    pub txid: TxId,
+
+    // List of all nullifiers spent in this Tx. These nullifiers belong to the wallet.
+    pub spent_sapling_nullifiers: Vec<zcash_primitives::sapling::Nullifier>,
+
+    // List of all nullifiers spent in this Tx. These nullifiers belong to the wallet.
+    pub spent_orchard_nullifiers: Vec<orchard::note::Nullifier>,
+
+    // List of all sapling notes received in this tx. Some of these might be change notes.
+    pub sapling_notes: Vec<SaplingNote>,
+
+    // List of all sapling notes received in this tx. Some of these might be change notes.
+    pub orchard_notes: Vec<OrchardNote>,
+
+    // List of all Utxos received in this Tx. Some of these might be change notes
+    pub transparent_notes: Vec<TransparentNote>,
+
+    // Total value of all the sapling nullifiers that were spent in this Tx
+    pub total_sapling_value_spent: u64,
+
+    // Total value of all the orchard nullifiers that were spent in this Tx
+    pub total_orchard_value_spent: u64,
+
+    // Total amount of transparent funds that belong to us that were spent in this Tx.
+    pub total_transparent_value_spent: u64,
+
+    // All outgoing sends
+    pub outgoing_tx_data: Vec<OutgoingTxData>,
+
+    // Whether this TxID was downloaded from the server and scanned for Memos
+    pub full_tx_scanned: bool,
+
+    // Price of Zec when this Tx was created
+    pub price: Option<f64>,
+}
+
+// set
+impl TransactionRecord {
+    pub fn new(status: ConfirmationStatus, datetime: u64, transaction_id: &TxId) -> Self {
+        TransactionRecord {
+            status,
+            datetime,
+            txid: *transaction_id,
+            spent_sapling_nullifiers: vec![],
+            spent_orchard_nullifiers: vec![],
+            sapling_notes: vec![],
+            orchard_notes: vec![],
+            transparent_notes: vec![],
+            total_transparent_value_spent: 0,
+            total_sapling_value_spent: 0,
+            total_orchard_value_spent: 0,
+            outgoing_tx_data: vec![],
+            full_tx_scanned: false,
+            price: None,
+        }
+    }
+    pub fn add_spent_nullifier(&mut self, nullifier: PoolNullifier, value: u64) {
+        match nullifier {
+            PoolNullifier::Sapling(sapling_nullifier) => {
+                self.spent_sapling_nullifiers.push(sapling_nullifier);
+                self.total_sapling_value_spent += value;
+            }
+            PoolNullifier::Orchard(orchard_nullifier) => {
+                self.spent_orchard_nullifiers.push(orchard_nullifier);
+                self.total_orchard_value_spent += value;
+            }
+        }
+    }
+}
+//get
+impl TransactionRecord {
+    pub fn get_transparent_value_spent(&self) -> u64 {
```

Comment:
What purpose does this fn serve, considering the field it returns is already public? 
Less is more when it comes to public API -- particularly when the API doesn't add significant value.

Reply:
i prefer a pattern where 0 means 0. total transparent_value_spent should be None until its filled in. however, this change will be a long way out.

Reply:
actually, this might be soon. because this value being unclear is causing inconsistencies

Reply:
actually, fair warning, im about to rewrite this entire structure